lib/widgets/actions.dart

```diff
@@ -16,101 +16,171 @@ import '../model/narrow.dart';
+    }
 
-  while (true) {
-    final result = await updateMessageFlagsForNarrow(connection,
-      anchor: anchor,
+    // Compare web's `mark_all_as_read` in web/src/unread_ops.js
+    // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    final didPass = await updateMessageFlagsStartingFromAnchor(
+      context: context,
+      // Include `is:unread` in the narrow.  That has a database index, so
+      // this can be an important optimization in narrows with a lot of history.
+      // The server applies the same optimization within the (deprecated)
+      // specialized endpoints for marking messages as read; see
+      // `do_mark_stream_messages_as_read` in `zulip:zerver/actions/message_flags.py`.
+      apiNarrow: narrow.apiEncode()..add(ApiNarrowIsUnread()),
+      // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
+      // will be the oldest non-muted unread message, which would
+      // result in muted unreads older than the first unread not
+      // being processed.
+      startingAnchor: AnchorCode.oldest,
       // [AnchorCode.oldest] is an anchor ID lower than any valid
       // message ID; and follow-up requests will have already
       // processed the anchor ID, so we just want this to be
       // unconditionally false.
       includeAnchor: false,
-      // There is an upper limit of 5000 messages per batch
-      // (numBefore + numAfter <= 5000) enforced on the server.
-      // See `update_message_flags_in_narrow` in zerver/views/message_flags.py .
-      // zulip-mobile uses `numAfter` of 5000, but web uses 1000
-      // for more responsive feedback. See zulip@f0d87fcf6.
-      numBefore: 0,
-      numAfter: 1000,
-      narrow: apiNarrow,
       op: UpdateMessageFlagsOp.add,
-      flag: MessageFlag.read);
-    if (!context.mounted) {
-      scaffoldMessenger.clearSnackBars();
-      return;
+      flag: MessageFlag.read,
+      onCompletedMessage: zulipLocalizations.markAsReadComplete,
+      progressMessage: zulipLocalizations.markAsReadInProgress,
+      onFailedTitle: zulipLocalizations.errorMarkAsReadFailedTitle);
+
+    if (!didPass || !context.mounted) return;
+    if (narrow is CombinedFeedNarrow && !useLegacy) {
+      PerAccountStoreWidget.of(context).unreads.handleAllMessagesReadSuccess();
     }
-    responseCount++;
-    updatedCount += result.updatedCount;
+  } catch (e) {
+    if (!context.mounted) return;
+    final zulipLocalizations = ZulipLocalizations.of(context);
+    await showErrorDialog(context: context,
+      title: zulipLocalizations.errorMarkAsReadFailedTitle,
+      message: e.toString()); // TODO(#741): extract user-facing message better
+    return;
+  }
+}
+
+Future<void> markNarrowAsUnreadFromMessage(
+  BuildContext context,
+  Message message,
+  Narrow narrow,
+) async {
+  final zulipLocalizations = ZulipLocalizations.of(context);
+  await updateMessageFlagsStartingFromAnchor(
```

Comment:
This won't work if the server's feature level is below 155. That's the point of the check in `markNarrowAsRead`, right?

To handle that:
* This function can just have an assert confirming the server is new enough (by checking `store.connection.zulipFeatureLevel`).
* The action sheet shouldn't offer the button if the server isn't new enough.
* At each of those places, leave a `// TODO(server-6)` comment so we can simplify the logic away when we drop support for affected versions. For discussion, see: https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#todo-server